core/src/main/scala/caliban/wrappers/Caching.scala

```diff
@@ -0,0 +1,288 @@
+              hint
+            case None       =>
+              _typeCache.put(key, None)
+              None
+          }
+      }
+    }
+
+    private def staticWrapper(state: Ref[CachePolicy]): ValidationWrapper[Any] = new ValidationWrapper[Any] {
+      override def wrap[R](
+        f: Document => ZIO[R, CalibanError.ValidationError, ExecutionRequest]
+      ): Document => ZIO[R, CalibanError.ValidationError, ExecutionRequest] = { (d: Document) =>
+        f(d).flatMap { request =>
+          def loop(policy: CachePolicy, field: Field): CachePolicy = {
+            val fieldHint = extractCacheDirective(field.directives)
+            val typeHint  = field.parentType.flatMap(cacheHintFromType)
+            val newPolicy = policy.restrict(fieldHint orElse typeHint)
+
+            field.fields.foldLeft(newPolicy)(loop)
+          }
+
+          val updated = loop(CachePolicy(CacheHint.default), request.field)
+
+          state.update(updated.merge).as(request)
+        }
+      }
+    }
+
+    private def fieldWrapper(state: Ref[CachePolicy]): FieldWrapper[Any] = new FieldWrapper[Any](false) {
+      override def wrap[R](
+        query: ZQuery[R, CalibanError.ExecutionError, ResponseValue],
+        info: FieldInfo
+      ): ZQuery[R, CalibanError.ExecutionError, ResponseValue] =
+        query.mapZIO { result =>
+          cacheOverride.get.flatMap {
+            case Some(overrideValue) => state.update(_.restrict(Some(overrideValue))) as result
+            case None                => ZIO.succeed(result)
+          }
+        }
+    }
+
+    private def overallWrapper(state: Ref[CachePolicy]): OverallWrapper[Any] = new OverallWrapper[Any] {
+      override def wrap[R](
+        f: GraphQLRequest => ZIO[R, Nothing, GraphQLResponse[CalibanError]]
+      ): GraphQLRequest => ZIO[R, Nothing, GraphQLResponse[CalibanError]] = { (request: GraphQLRequest) =>
+        (f(request) zipWith state.get) { (response, cacheState) =>
+          val cacheControl = cacheState.hint
+          val maxAge       = cacheControl.maxAge getOrElse settings.defaultMaxAge
+          val scope        = cacheControl.scope getOrElse settings.defaultScope
+          response.copy(
+            extensions = Some(
+              ResponseValue.ObjectValue(
+                response.extensions.foldLeft[List[(String, ResponseValue)]](
+                  List(
+                    DirectiveName -> ResponseValue.ObjectValue(
+                      List(
+                        "version" -> IntValue(2),
+                        "policy"  -> ResponseValue.ObjectValue(
+                          List(
+                            MaxAgeName -> IntValue(maxAge.toSeconds),
+                            ScopeName  -> StringValue(scope.toString)
+                          )
+                        )
+                      )
+                    )
+                  )
+                )((s, extensions) => extensions.fields ++ s)
+              )
```

Comment:
Very nitpick (sorry 😅): I think this is a bit more performant / readable

```suggestion
              ResponseValue.ObjectValue(
                (DirectiveName -> ResponseValue.ObjectValue(
                  List(
                    "version" -> IntValue(2),
                    "policy" -> ResponseValue.ObjectValue(
                      List(
                        MaxAgeName -> IntValue(maxAge.toSeconds),
                        ScopeName -> StringValue(scope.toString)
                      )
                    )
                  )
                )) :: response.extensions.map(_.fields).getOrElse(Nil)
              )
```

Reply:
Done! But with `fold` so we don't need an intermediate `Option` construction